ads/google/a4a/performance.js

```diff
@@ -194,36 +198,65 @@ export class GoogleAdLifecycleReporter extends BaseLifecycleReporter {
    * @override
    */
   sendPing(name) {
-    this.emitPing_(this.buildPingAddress_(name));
+    this.emitPing_(this.buildPingAddress_(name, this.extraVariables_));
   }
 
   /**
+   * @param {string} variable
+   * @param {string|number} value
+   * @override
+   */
+  setPingVariable(variable, value) {
+    this.extraVariables_[variable] = value;
+  }
+
+  /**
+   * @return {number}
+   */
+  getSlotId() {
+    return this.slotId_;
+  }
+
+  /** @override */
+  reset() {
+    this.extraVariables_ = new Object(null);
+  }
+
+
+  /**
    * @param {string} name  Metric name to send.
+   * @param {Object<string, string|number>=} opt_extraParams
    * @returns {string}  URL to send metrics to.
    * @private
    */
-  buildPingAddress_(name) {
+  buildPingAddress_(name, opt_extraParams) {
     const stageId = LIFECYCLE_STAGES[name] || 9999;
     const delta = Date.now() - this.initTime_;
-    // Note: QQid comes from a network header and eid could, potentially, be
-    // injected by a publisher.  Treat both of them as unverified user content
-    // and encode before inserting them into URI.
-    const encodedQqid = this.qqid_ ?
-        encodeURIComponent(this.qqid_) : false;
-    const qqidParam = encodedQqid ?
-        `&qqid.${this.slotId_}=${encodedQqid}` : '';
-    const eid = this.element_.getAttribute(EXPERIMENT_ATTRIBUTE);
-    const eidParam = eid ? `&e=${encodeURIComponent(eid)}` : '';
+    let extraParams = '';
+    if (opt_extraParams) {
+      const paramList = [];
+      for (const param in opt_extraParams) {
+        if (!!param) {
+          paramList.push(
+              encodeURIComponent(param) + '=' +
+              encodeURIComponent(opt_extraParams[param]));
+        }
+      }
+      if (paramList.length > 0) {
+        extraParams = '&' +
+            this.urlReplacer_.expandStringSync(paramList.join('&'));
```

Comment:
Sync vs async: I didn't want the ping to block trying to get some bit of info that might take a while.  This way, any such requests just get dropped.  But if you think the extra params are worth it, we can make it async.  (Still won't get us scroll offset, which is an analytics param.)

Key could be expanded: Ooh.  Good point.  I'm just reluctant to call expand a bunch of different times.  Looking at it, it's not clear how much overhead there is per call, but it's non-zero.  Currently, expansion is typically applied to whole URLs and there's an "on your own head be it" view of people putting expansion variables in unexpected places.